distribution/docker/debian/Dockerfile

```diff
@@ -1,22 +1,55 @@
-FROM docker.io/debian:bullseye-slim AS builder
-
-WORKDIR /vector
-
-COPY vector_*.deb ./
-RUN dpkg -i vector_*_"$(dpkg --print-architecture)".deb
-
-RUN mkdir -p /var/lib/vector
-
-FROM docker.io/debian:bullseye-slim
-
-RUN apt-get update && apt-get install -y --no-install-recommends ca-certificates tzdata systemd && rm -rf /var/lib/apt/lists/*
-
-COPY --from=builder /usr/bin/vector /usr/bin/vector
-COPY --from=builder /usr/share/doc/vector /usr/share/doc/vector
-COPY --from=builder /etc/vector /etc/vector
-COPY --from=builder /var/lib/vector /var/lib/vector
-
-# Smoke test
-RUN ["vector", "--version"]
-
-ENTRYPOINT ["/usr/bin/vector"]
+ARG ALPINE_IMAGE=docker.io/alpine:3.15
+ARG DEBIAN_IMAGE=docker.io/debian:bullseye-slim
+
+#-------------------------------------------------------------------------------
+FROM ${ALPINE_IMAGE} AS staging
+
+# Set a well-known working directory for staging
+WORKDIR /staging
+
+# Prepare a staging root filesystem
+RUN mkdir -p \
+  rootfs/etc/vector \
+  rootfs/usr/bin \
+  rootfs/var/lib/vector
+
+COPY docker-entrypoint.sh rootfs/usr/bin/docker-entrypoint.sh
+COPY --from=docker.io/tianon/gosu:latest gosu rootfs/usr/bin/gosu
+
+# Copy and unpack Vector distribution
+COPY vector-*-unknown-linux-gnu*.tar.gz .
+RUN mkdir vector \
+    && tar zxvf vector-*-"$(cat /etc/apk/arch)"-unknown-linux-gnu*.tar.gz \
+    -C vector --strip-components=2
+
+# Prepare Vector in the staging root filesystem
+RUN cp vector/bin/* rootfs/usr/bin/ \
+  && cp -r vector/config/* rootfs/etc/vector/
+
+#-------------------------------------------------------------------------------
+FROM ${DEBIAN_IMAGE}
+
+# Set image labels
+ARG VECTOR_VERSION
+LABEL org.opencontainers.image.source="https://github.com/vectordotdev/vector" \
+  org.opencontainers.image.title="Vector (Debian variant)" \
+  org.opencontainers.image.url="https://vector.dev/" \
+  org.opencontainers.image.version="${VECTOR_VERSION}"
+
+# Copy root filesystem from staging
+COPY --from=staging /staging/rootfs/ /
+
+RUN apt-get update \
+    && apt-get upgrade -y \
+    && apt-get install -y --no-install-recommends \
+        ca-certificates \
+        tini \
+        tzdata \
+    && rm -rf /var/lib/apt/lists/* \
+    && useradd --system \
+        --shell /sbin/nologin \
+        --user-group vector
```

Comment:
Looking at https://github.com/docker-library/postgres/blob/master/15/bullseye/Dockerfile and https://github.com/docker-library/mongo/blob/1d2a1b08bd24135e9ce8ecec55b27cfc5fb6f0a1/5.0/Dockerfile#L9-L14 it seems like setting a static uid/gid of 999 is popular. Do you have thoughts on that?

Reply:
Funny you say that - that also seems to happen when we're not specifying it, we can certainly choose an ID but we gotta pick a good number

```shell
❯ docker top vector
UID                 PID                 PPID                C                   STIME               TTY                 TIME                CMD
root                483519              483493              0                   21:40               ?                   00:00:00            /usr/bin/qemu-x86_64 /usr/bin/tini tini -- docker-entrypoint.sh vector
>> 999 <<           483551              483519              8                   21:40               ?                   00:00:00            /usr/bin/qemu-x86_64 /usr/bin/vector vector
```

Reply:
Another aspect is if the UID happens to match a user known to the OS you get known by funny things, like the alpine variant shows UID as `_apt`

Reply:
I would also recommend using a static (known) explicit UID/GID.
That you are getting `999` is just casual when you don't specify it.
Nothing guarantees you will get the same assignment in the future.

Reply:
Ah, I guess outstanding item here @jszwedko is setting a static UID/GID if desired (and picking it)

Reply:
This matches the behavior of the apt/rpm packages so I'm going to leave as is and we can revisit for all platforms.